### PR TITLE
GitHub Like Diff

### PR DESCRIPTION
**Pull request by sotarok**
Thank you so much @kaz @b4b4r07  

## Spec

- Diff like GitHub
- Show diff even if it is the first revision

image

## Also

- Close #170 
- Close #171 
- Close #173

### CODE REVIEW

**Comment by suzuki** [comment]
I found different points from the behavior of the past version.

- "View diff" link
    - Can't toggle
    - The icon did not change
- 1st revision
    - Always open

File: resource/css/crowi.scss [hunk 1]
```diff
@@ -251,6 +251,11 @@ footer {
   border-radius: 3px;
 }
 
+// adust
```
Comment (suzuki) [inline resource/css/crowi.scss]: typo ?

File: resource/css/crowi.scss [hunk 2]
```diff
@@ -251,6 +251,11 @@ footer {
...
+.d2h-file-header {
```
Comment (suzuki) [inline resource/css/crowi.scss]: [memo] class for "diff2html", maybe.

File: resource/js/app.js [hunk 3]
```diff
@@ -21,10 +22,13 @@ crowi.fetchUsers();
 const crowiRenderer = new CrowiRenderer();
 window.crowiRenderer = crowiRenderer;
 
+const pageId = $('#content-main').data('page-id');
```
Comment (suzuki) [inline resource/js/app.js]: We can use `document.querySelector('#content-main')`
see: https://github.com/crowi/crowi/pull/174/files#diff-39992ee08af86834f2b5095ded60c098R25

File: resource/js/app.js [hunk 4]
```diff
@@ -35,3 +39,8 @@ Object.keys(componentMappings).forEach((key) => {
     ReactDOM.render(componentMappings[key], elem);
   }
 });
+
+// うわーもうー
+$('a[data-toggle="tab"][href="#revision-history"]').on('show.bs.tab', function() {
```
Comment (suzuki) [inline resource/js/app.js]: We can use `document.querySelector('a[data-toggle="tab"][href="#revision-history"]')`.
But I did not test this event using it.

File: resource/js/components/Common/Icon.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    return (
+      <i className={"fa fa-" + name} />
+    );
+  }
+}
+
+Icon.propTypes = {
+  name: React.PropTypes.string,
```
Comment (suzuki) [inline resource/js/components/Common/Icon.js]: We can use `isRequired` here.
I think the Font Awesome requires name.

File: resource/js/components/PageHistory.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+import Icon from './Common/Icon';
+import PageRevisionList from './PageHistory/PageRevisionList';
+
+export default class PageHistory extends React.Component {
+
+  constructor(props) {
+    super(props);
+
+    this.crowi = window.crowi; // FIXME
```
Comment (suzuki) [inline resource/js/components/PageHistory.js]: How about changing `crowi` to `props`?
see: https://github.com/crowi/crowi/pull/174/files#diff-39992ee08af86834f2b5095ded60c098R37

File: resource/js/components/PageHistory.js [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+      }
+
+      cursor = revision;
+    }
+
+    return cursor;
+  }
+
+  onDiffOpenClicked(revision)
+  {
```
Comment (suzuki) [inline resource/js/components/PageHistory.js]: We need to select (decide) the coding style.

File: resource/js/components/PageHistory/Revision.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    return (
+      <div className="revision-history-main">
+        {pic}
+        <div className="revision-history-author">
+          <strong>{author.username}</strong>
+        </div>
+        <div className="revision-history-meta">
+         <p>
```
Comment (suzuki) [inline resource/js/components/PageHistory/Revision.js]: indent

**Comment by sotarok** [comment]
> "View diff" link

I changed the spec.
To simplize the behavior, i removed the feature of closing diff.
On my mine, 
(It's like "Expand codes around the diff" of GItHub's diff. We can't close the opened diff.)

image



> 1st  revision

I also changed the spec.
Because, the first virsion, in my experience, I used to read the original contents of the page.

File: resource/js/components/PageHistory.js (hunk 6 above)
Comment (sotarok) [inline resource/js/components/PageHistory.js]: Ah, nice

File: resource/css/crowi.scss (hunk 1 above)
Comment (sotarok) [inline resource/css/crowi.scss]: :stuck_out_tongue_closed_eyes:

**Pull request merged**